kotlinx-coroutines-test/common/src/TestCoroutineDispatchers.kt

```diff
@@ -15,10 +17,32 @@ import kotlin.coroutines.*
  * This dispatcher is similar to [Dispatchers.Unconfined]: the tasks that it executes are not confined to any particular
  * thread and form an event loop; it's different in that it skips delays, as all [TestDispatcher]s do.
  *
+ * Like [Dispatchers.Unconfined], this one does not provide guarantees about the execution order when several coroutines
+ * are queued in this dispatcher. However, we ensure that the [launch] and [async] blocks at the top level of [runTest]
+ * are entered eagerly. This allows launching child coroutines and not calling [runCurrent] for them to start executing.
+ *
+ * ```
+ * @Test
+ * fun testEagerlyEnteringChildCoroutines() = runTest(UnconfinedTestDispatcher()) {
```

Comment:
By the way, what do you think about providing `runTestUnconfined` for a better migration path?

Reply:
I don't like it because of this confusing behavior:
```kotlin
val ctx = StandardTestDispatcher()

@Test
fun test() = runTestUnconfined(ctx) {
  // not unconfined
}
```